Index: site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java
===================================================================
--- site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java	(revision 104522)
+++ site-manage-tool/tool/src/java/org/sakaiproject/site/tool/SiteAction.java	(working copy)
@@ -21,11 +21,15 @@
 
 import java.io.IOException;
 import java.lang.reflect.Method;
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Calendar;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.Date;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Hashtable;
@@ -181,6 +185,9 @@
 	private static org.sakaiproject.sitemanage.api.model.SiteSetupQuestionService questionService = (org.sakaiproject.sitemanage.api.model.SiteSetupQuestionService) ComponentManager
 	.get(org.sakaiproject.sitemanage.api.model.SiteSetupQuestionService.class);
 	
+	private static org.sakaiproject.email.api.EmailService emailService = (org.sakaiproject.email.api.EmailService) ComponentManager
+	.get(org.sakaiproject.email.api.EmailService.class);
+	
 	private static final String SITE_MODE_SITESETUP = "sitesetup";
 
 	private static final String SITE_MODE_SITEINFO = "siteinfo";
@@ -430,7 +437,18 @@
 	private static final String STATE_SITE_ACCESS_PUBLISH = "state_site_access_publish";
 	
 	private static final String STATE_SITE_ACCESS_INCLUDE = "state_site_access_include";
+	
+	/** Shopping Period state */
+	private static final String STATE_SITE_ACCESS_SHOPPING_VISIBILITY = "state_site_access_visibility";
 
+	private static final String STATE_SITE_ACCESS_SHOPPING_VISIBILITY_START = "state_site_access_shopping_start";
+
+	private static final String STATE_SITE_ACCESS_SHOPPING_VISIBILITY_END = "state_site_access_shopping_end";
+	
+	private static final String STATE_SITE_ACCESS_SHOPPING_ROLE = "state_site_access_shopping_role";
+	
+	private static final String STATE_SITE_ACCESS_SHOPPING_REALM = "state_site_access_shopping_realm";
+
 	/** the list of selected user */
 	private static final String STATE_SELECTED_USER_LIST = "state_selected_user_list";
 
@@ -967,10 +985,8 @@
 				}
 	
 				state.setAttribute(STATE_SITE_MODE, site_mode);
-			}
+		}
 
-
-		
 	} // initState
 	
 	
@@ -2298,6 +2314,19 @@
 				}
 				context.put("published", state.getAttribute(STATE_SITE_ACCESS_PUBLISH));
 				context.put("include", state.getAttribute(STATE_SITE_ACCESS_INCLUDE));
+				context.put("shopping_visibility", state.getAttribute(STATE_SITE_ACCESS_SHOPPING_VISIBILITY));
+				String start = (String) state.getAttribute(STATE_SITE_ACCESS_SHOPPING_VISIBILITY_START);
+				if (start == null) {
+					start = "";
+				}
+				String end = (String) state.getAttribute(STATE_SITE_ACCESS_SHOPPING_VISIBILITY_END);
+				if (end == null) {
+					end = "";
+				}
+				context.put("shopping_visibility_start", start);
+				context.put("shopping_visibility_end", end);
+				context.put("all_roles", site.getRoles());
+				context.put("shoppingPeriodInstructorEditable", ServerConfigurationService.getBoolean("delegatedaccess.shopping.instructorEditable", false));
 
 				if (siteType != null && !unJoinableSiteTypes.contains(siteType)) {
 					// site can be set as joinable
@@ -6571,7 +6600,7 @@
 					}
 				}
 			}
-			state.setAttribute(STATE_JOINERROLE, joinerRole); 
+			state.setAttribute(STATE_JOINERROLE, joinerRole);
 		}
 		catch (Exception e)
 		{
@@ -7213,6 +7242,9 @@
 		readInputAndUpdateStateVariable(state, params, "include", STATE_SITE_ACCESS_INCLUDE, true);
 		readInputAndUpdateStateVariable(state, params, "joinable", STATE_JOINABLE, true);
 		readInputAndUpdateStateVariable(state, params, "joinerRole", STATE_JOINERROLE, false);
+		readInputAndUpdateStateVariable(state, params, "shoppingVisibility",      STATE_SITE_ACCESS_SHOPPING_VISIBILITY, false);
+		readInputAndUpdateStateVariable(state, params, "shoppingVisibilityStart", STATE_SITE_ACCESS_SHOPPING_VISIBILITY_START, false);
+		readInputAndUpdateStateVariable(state, params, "shoppingVisibilityEnd",   STATE_SITE_ACCESS_SHOPPING_VISIBILITY_END, false);
 		
 		boolean publishUnpublish = state.getAttribute(STATE_SITE_ACCESS_PUBLISH) != null ? ((Boolean) state.getAttribute(STATE_SITE_ACCESS_PUBLISH)).booleanValue() : false;
 		
Index: site-manage-tool/tool/src/bundle/sitesetupgeneric.properties
===================================================================
--- site-manage-tool/tool/src/bundle/sitesetupgeneric.properties	(revision 104522)
+++ site-manage-tool/tool/src/bundle/sitesetupgeneric.properties	(working copy)
@@ -1057,3 +1057,14 @@
 sitegen.siteinfolist.usermodify = Modified by
 
 useOfficialDescription=Use Official Description
+
+## Site Shopping Period
+shopping.header = Shopping Period Access
+shopping.start.label = Start Date
+shopping.end.label = End Date
+shopping.visibility.label = Visibility
+shopping.role.label = Role
+shopping.role.instructions = Shoppers have the same access as this realm/role.
+shopping.instructions=You have the ability to add, update, or remove your site's shopping access.  To add or update, just fill in the fields below.  To remove, clear all fields and set all options to none.
+shopping.noauth=none
+shopping.showTools.label=Show Tools
Index: site-manage-tool/tool/src/webapp/vm/sitesetup/chef_site-siteInfo-editAccess.vm
===================================================================
--- site-manage-tool/tool/src/webapp/vm/sitesetup/chef_site-siteInfo-editAccess.vm	(revision 104522)
+++ site-manage-tool/tool/src/webapp/vm/sitesetup/chef_site-siteInfo-editAccess.vm	(working copy)
@@ -4,6 +4,9 @@
 <script type="text/javascript" src="/library/js/jquery/jquery-ui/js/jquery-ui.js"></script>
 <link type="text/css" href="/library/js/jquery/jquery-ui/css/smoothness/jquery-ui.css" rel="stylesheet" media="screen" />
 <script type="text/javascript" src="/sakai-site-manage-tool/js/site-manage.js"></script>
+#if ($shoppingPeriodInstructorEditable)
+	<script type="text/javascript" src="/sakai-site-manage-tool/js/site-manage-shopping.js"></script>
+#end
 <script type="text/javascript">
 <!-- hide from non-JS browsers
 	function checkPublish(checked)
@@ -42,6 +45,8 @@
 			utils.resizeFrame('shrink');
 		}
 	}
+	
+	
 // -->	
 </script>
 <link type="text/css" href="/sakai-site-manage-tool/css/site-manage.css" rel="stylesheet" media="screen" />
@@ -212,12 +217,77 @@
 			</p>
 			<p/>
 		</div>
+
+		<span id="siteId" style="display:none">$site.id</span>
+		#if ($shoppingPeriodInstructorEditable)
+		<div id="shoppingVisibilityDiv" style="display:none">
+			<!-- easy access for JS -->
+			
+			<h4>$tlang.getString("shopping.header")</h4>
+			<p class="instruction">
+				$tlang.getString("shopping.instructions")
+			</p>
+			<div class="indnt1">
+				<table>
+				<tr>
+				<td align="right">
+			 	<label for="shoppingVisibility">$tlang.getString("shopping.visibility.label"): </label>
+				</td>
+				<td>
+				<select name="shoppingVisibility" id="shoppingVisibility">
+					<option value="">$tlang.getString("shopping.noauth")</option>
+				</select>
+				</td>
+				</tr>
+				<tr>
+                                <td align="right">
+				<label for="shoppingVisibilityStart">$tlang.getString("shopping.start.label"): </label>
+				</td>
+				<td>
+				<input type="text" id="shoppingVisibilityStart" name="shoppingVisibilityStart" value="$shopping_visibility_start"/>
+				</td>
+				</tr>
+                                <tr>
+                                <td align="right">
+				<label for="shoppingVisibilityEnd">$tlang.getString("shopping.end.label"): </label>
+				</td>
+                                <td>
+				<input type="text" id="shoppingVisibilityEnd" name="shoppingVisibilityEnd" value="$shopping_visibility_end"/>
+				</td>
+				</tr>
+				<tr>
+				<td align="right">
+				<label for="shoppingRole">$tlang.getString("shopping.role.label"): </label>
+				</td>
+				<td>
+				<select name="shoppingRole" id="shoppingRole">
+					<option value="">$tlang.getString("shopping.noauth")</option>
+				</select>
+				</td>
+				</tr>
+				<tr>
+					<td>
+						<label for="showTools">$tlang.getString("shopping.showTools.label"): </label>
+					</td>
+					<td>
+						<select name="showTools" id="showTools" multiple="multiple" style="height: 100px">
+							<option value=""></option>
+						</select>
+					</td>
+				</tr>
+				</table>
+			</div>
+			</div>
+
+		</div>
+		#end
 		<input type="hidden" name="continue" value="$!continue" />
 		<input type="hidden" name="back" value="$!backIndex" />
 		<input type="hidden" name="templateIndex" value="$!templateIndex" />
 		<div class ="act">
 			#if($!site)
-				<input type="submit" accesskey="s" class="active" name="eventSubmit_doUpdate_site_access" value="$tlang.getString("ediacc.upd")" />
+				<input type="submit" accesskey="s" id="updateButton" 
+					class="active" name="eventSubmit_doUpdate_site_access" value="$tlang.getString("ediacc.upd")" />
 				<input type="submit" accesskey="x"  name="eventSubmit_doCancel" value="$tlang.getString("ediacc.bac")" />
 #else
 				<input type="submit" accesskey="s" class="active" name="eventSubmit_doUpdate_site_access" value="$tlang.getString("ediacc.con")" />
